### ISSUE TITLE
ASB Metrics

### ISSUE THREAD

**Issue by bonomat** [comit-network/xmr-btc-swap#448]
As a maker I would like a cli command to see for all swaps: 

- when they were started
- amounts being swapped (incl rate at start time)
- date at each step (fund/redeem/refund) including rate
- taker's id

Hint: currently old states are overwritten in the DB. We could just add a new DB entry for each event including details.

**Comment by thomaseizinger** [comment]
Rust-libp2p is exploring OpenMetrics: https://github.com/libp2p/rust-libp2p/pull/2063

Don't have much experience but this like a useful standard that we could follow here where we don't have to reinvent the wheel and make it easy for the user to later tap into a wider ecosystem of monitoring.

**Comment by bonomat** [comment]
OpenMetrics (same for [Prometheus](https://prometheus.io/docs/concepts/metric_types/) does not seem to be suitable for our needs. 

It is designed to monitor services from a _data_ perspective: 

> "Metrics are a specific kind of telemetry data. They represent a snapshot of the current state for a set of data. "

Examples range from CPU utilization and I/O load to service response times. 

The continue with : 

> "They are distinct from logs or events, which focus on records or information about individual events." 

[Source](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#overview)

The latter one is more what we need, i.e. record and collect events.

It gets more clear that OpenMetrics does not fit when reading through the defined [metric types](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#metric-types).

**Comment by bonomat** [comment]
I propose to follow a store-and-log approach instead: 

1) store interesting events in the db (create a new `tree` for `metrics`).
2) add a new command which reads all (or only events for a specific swap-id) and prints it to stdout.
  a) this allows us to decide the format for the events later-on.

**Comment by bonomat** [comment]
What I would like to see (the format of the log does not matter atm): 

```
Metric: 
{
	`swap-id`: `Id`,
	'states` : `[State]`, 
	`btc-amount`: `Number`,
	`xmr-amount`: `Number`,
	`counter-party`: `PerrId`
}

State: 
{
    `recorded`: `Timestamp` // when this state was recorded
	`rate`: `Number` // what was the exchange rate at this point
	`tx-id`: `TxId` // if any
}

```

**Comment by da-kami** [comment]
In general:

Yeah, what you describe has notthing to do with `Metrics` but more with detailed `Events` - once we aggregate these Events and analyze them we have actual `Metrics`.

To be more concrete: What you outline above as `Metric` should be named `Swaps` of `SwapDetails` I think:

```
Swaps: 
{
	`swap-id`: `Id`,
	'states` : `[State]`, 
	`btc-amount`: `Number`,
	`xmr-amount`: `Number`,
	`counter-party`: `PerrId`
}
```

Only once we analyze multiple of these swaps and we would actually have Metrics as in "5 out of 10 swaps finished successfully with state ...". 

I am fine with recording more swap details in the Sled DB in separate trees for a "quick solution" - but if we plan for the long run it would be better to do what I outline below.

---

I had the alternative idea to improve the logs and let an external software handle our "Metrics" through the logs. We already use swap and peer-id contexts within the logging and could easily extract swap / peer information using log frameworks such as [vector](https://github.com/timberio/vector)

Advantages: 
* We don't clutter our application code with recording events that don't serve any purpose other than log-details
* We improve our logs rather than adding additional "events" on top
* We can easily visualize data (UI) by pumping it into other software - Vector supports multiple `sinks`
*  We can define transformation rules without having to touch application code and release...


Disadvantage: 
* We have to set it up and get into Vector. (it looks fairly straight forward, I started playing with it but don't have a working example yet...)
* It does not work "out of the box" - i.e. other ASB providers would have to set e.g. Vector up themselves. We could provide documentation.

**Comment by thomaseizinger** [comment]
Sad that OpenMetrics doesn't support more complex things that than. The big advantage would have been that we don't need to store this data. An ever-growing database is not particularly ideal and metrics / logs are things that you usually don't care about past a certain point in time.

If we store things in a database, can we use this an opportunity to start migrating to SQLite? The actual extraction of metrics could then be as simple as loading the database into any SQL tool and running a couple of queries against it.

That should reduce the required development effort significantly. Also, if we create a separate reporting database, deleting that one is safe if the user ever wants to clean up storage space.

**Comment by bonomat** [comment]
> Yeah, what you describe has nothing to do with `Metrics` but more with `Events` and just swap-state - once we aggregate these Events and analyze them we have actual `Metrics`.

That's a good summary :)

I had a quick look into Vector as well. It's a _log analyzer_: 

> Vector is a high-performance observability data pipeline that allows you to collect, transform, and route all your logs and metrics.

It even has an export feature to send data to prometheus (which afaik is in OpenMetrics format).

If we want to go this way, we can ignore the DB and print more details on the logs. 

@thomaseizinger : what do you think?

**Comment by da-kami** [comment]
> > Yeah, what you describe has nothing to do with `Metrics` but more with `Events` and just swap-state - once we aggregate these Events and analyze them we have actual `Metrics`.
> 
> That's a good summary :)
> 
> I had a quick look into Vector as well. It's a _log analyzer_, maybe comparable to LogStash:
> 
> > Vector is a high-performance observability data pipeline that allows you to collect, transform, and route all your logs and metrics.
> 
> It even has an export feature to send data to prometheus (which afaik is in OpenMetrics format).
> 
> If we want to go this way, we can ignore the DB and print more details on the logs.
> 
> @thomaseizinger : what do you think?

I think it boils down to either refactoring the DB to be relational and then use some tools on top of the relational DB for analyzis OR ignore the DB and go with the logs. I am not sure what is the better approach, but I think the log approach would give us faster results. 
Refactoring the DB would be a good thing though, because that is tech dept that we keep carrying around...

**Comment by bonomat** [comment]
After quick chat with @da-kami: 

1. Analyzing and coming up with Metrics for the ASB is in the role of the ASB provider and hence we can see this as a nice-to-have feature. 
2. It is hard to get the metrics _right_ for everyone hence we should strive for a flexible solution.
3. A new DB (or sled-tree) would add additional complexity because we would need to think of an upgradable or flexible data-schema. Additionally, it was mentioned that SQL should replace Sled eventually. We should not jump the gun on this topic because we want some more information to analyze.

Because of 1. + 2. it would be easier to add _some more_ information into our logs. These can then be analyzed using tools like [Vector](https://vector.dev/) or [LogStash](https://www.elastic.co/logstash)+Elastic or other tools.
